### DISCUSSION TITLE
Handle Registration/Link via Dbus?

### DISCUSSION THREAD

**Discussion by bublath** [AsamK/signal-cli#520] [Ideas]
I digged a bit thru the code to potentially implement a registration via Dbus. However there seems to be a hen-and-egg problem here. 
Starting the application in Dbus mode requires a valid registration and getting a RegistrationManager from a Manager is also not straight forward.
Having registration via Dbus would create the possibility to guide users conveniently through the registration process, even potentially embedding the Captcha and redirecting the Captcha by creating a protocal handler to catch the signalcaptcha:// code.
Same for the "link" where the qrcode could be put on the screen.
All this would require to start signal-cli in an unregistered mode, so it would listen to dbus commands (but typically reply with a kind "not registered" exception). What would be the complexity of this?

**Comment by bublath** [comment]
@AsamK : Before digging into the code more, can this work:
- When starting up, check for valid registry
- If not present in daemon mode, instanciate RegistrationManager instead of Manager (or ProvisioningManager - can we merge them?)
- While not registered, block all dbus commands that require it, basically we only allow isRegistered, register, verify, captcha, link and setNumber
- I would prefer to even set the number dynamically, since that will allow a generic daemon installation (more below)
- Once registration is complete, either simply exit (systemd will restart it automatically) or delete the old manager and create the normal one

The beauty of this would be to provide a standard installation (e.g. a Debian package) that installs a generic signal-cli daemon). That would increase usability significantly. It would also take care about the platform dependency (provide x64 and arm packages) and package dependencies (e.g. the right Java version).
The client programs can then use the methods to guide the users thru the registration process in a user friendly way, which includes guiding them to the Captcha page or displaying the qrcode
By being able to set the number, it would even be possible to switch between profiles on-the-fly and register additional numbers (not sure about the use case, but still nice)

Sounds feasible?

**Reply by AsamK** [reply]
Interesting use case. signal-cli now already supports starting the daemon without a user name, in which case it will expose all registered accounts a separate dbus objects.
That case could be extended to also start the daemon if no registered accounts are present, in which case it would expose a new dbus object with a new dbus interface (maybe `org.asamk.SignalInit`?), that provides the register and link commands.

**Reply by bublath** [reply]
How about this: Always provide the standard path interface /org/asamk/Signal - even in the multiple accounts case.
A client would always use this to connect first and call isRegistered() to findout if a registration is necessary.
A second method List<String> listRegistered() will inform the client if there is one or multiple accounts. It can then potentially offer a choice to the user (and connect to the specific path) or if its just one account (or an old client that does not know anything about these options) will just continue using the default one.

Even in the multi account case, the default interface could continue to exist, to offer registration/linking of additional devices.
For the implementation, do I get this right that in case of DaemonCommand simply n threads are started into the background and the main one just exits?
So in App.java ~line 136  if (usernames.size() == 0 && !useDbus && !useDbusSystem) {
to avoid the "No local users" error.
We should end up in handleMultiLocalCommand and would add 
        if (usernames.size()==0) {
            managers.add(new GenericManager(dataPath, serviceEnvironment));
        }
since the for loop would do nothing (no usernames).
then handleCommand() will call DaemonCommand with our new Manager

The new manager will then provide the necessary infrastructure to accept and execute the required registration and link commands.
Thread run() would still need to check on the type of manager and for our new Manager export differently something like:
conn.exportObject(objectPath, new DbusSignalInit(m));

In this new Interface we would implement the interface for the registrations. We will probably also need "dummy" overrides for everything in DbusSignalImpl that throw something like NotRegisteredException()? Since many call the manager, this will probably be more transparent then adding dummy methods in the manager.

I have to admit I'm no Ace Java Programmer, so that will be quite a challenge for me and before I start anything, I would like to know if my concept is on the right track.

**Reply by AsamK** [reply]
In multi account mode, clients can get the list of available accounts via dbus introspection, to see which object paths are available.

I've had a go at this and implemented a new org.asamk.SignalControl dbus interface on the default /org/asamk/Signal path, which provides register/link/verify methods. The code is on a separate branch, mostly untested and not completely finished, especially error handling needs to be improved.

What do you think about the interface?
https://github.com/AsamK/signal-cli/commit/599825e9b9abc5d034e065bd72d3ccac84cb0bd7

**Reply by bublath** [reply]
Looking at the interface, I would probably add a new **RequiresCaptchaException()** to make it easier for the client to ask the user to retry with a Captcha.
To avoid clients to run into an error, we should override the **isRegistered()** method with return "false". It would basically be the first mandatory step for any client to find out what he is dealing with. Also still provide the **version()** method so a client can check already in this stage which version he is dealing with.

So if I get your commits plus these patches right, you will go into the DaemonCommand with an empty list of managers, but with your extra SignalCreator.
I struggle a bit to understand how you deal with having no username in that lamda-function you put. How will the m.getUsername() work?

I just assume that after running this without -u and empty config directory, I can access the new interface to start registering/linking on the standard base path (/org/asamk/Signal).
So if a client tries to call any standard methods, he would probably already get an error from the Dbus layer, since these methods are not exposed, correct? With isRegistered() that should be ok - the client can simply avoid this.

If that works alright this would be a big step ahead to a generic installation via .deb packages and improved user experience by letting the client guide through the registration.

Something more to consider (but maybe as a next step):
- Even provide the registration/link interface in normal operation mode to add additional accounts.
- Always provide the base path (even in multi account mode) for further registrations etc. and doing a query on the list of registered accounts. This is to avoid clients to fall into the trap of not finding an interface once a second account was added. It would always be able to rely on the base path to be there (check the version, registered...) and go from there (let the user choose an existing account and connect to that interface).

**Reply by bublath** [reply]
Maybe to add some thoughts on implementation options:
When I looked into it, I was considering to create a generic Manager class, that provides all basic methods and does not require a username. The current manager class as well as RegistrationManager and ProvisionManager could be derived from this. This could help to streamline the Manager class a bit (which has grown quite large) - maybe even the derived "Main Manager" could be split in several use cases. By this all the command handlers, that require a Manager class, could basically remain unchanged (maybe some additional checks) while the generic Manager would probably need to implement some dummy methods that would be overridden in the specific managers.

I get that your approach is less intrusive and probably easier to do, since what I propose here is a major refactoring on the Manager and requires more analysis on what functionality goes where. 

On your remark above on Dbus inspection: While that of course would be a possibility, it would require the clients to implement quite a bit of additional code to get and parse that data. That's why I propose to always provide the base Path with some methods to detect everything. As a client developer, I would probably not invest into that inspection, but rather try the default path and on error assume that signal-cli runs in multi-mode otherwise.

**Reply by bublath** [reply]
I checked out your branch and tried it a bit.
I could start with empty config and without username. Reading the introspective worked.
Then I tried "link" (the easiest case to test).
` DEBUG org.freedesktop.dbus.messages.MethodReturn - Appending sig: s data: [tsdevice:/?uuid=(deleted)]
[main] DEBUG org.freedesktop.dbus.connections.transports.UnixSocketTransport - Disconnecting Transport
[DBus Sender Thread-1] DEBUG org.freedesktop.dbus.connections.impl.DBusConnection - Exception while sending message.
org.freedesktop.dbus.exceptions.NotConnected: Disconnected
`
So it looks like it got the message and got the link code. I pasted the data from the log into qrencode and tried to link. Signal App accepted the code, but after confirming said "no device" - I guess there would be a handshake and since signal-cli ran into errors this could not be done.

Let me know when you have time to continue, and I will further extend my client to use the new interfaces in order to test it.

**Reply by AsamK** [reply]
Thanks for the feedback. I've pushed another commit, that adds a `version` and `listAccounts` method to the control interface. An `isRegistered` method is not necessary, as it's a separate dbus interface.
The idea is that in multi account mode the `SignalControl` interface is exported at the base path to allow registering/linking new accounts. The account are then exported as separate objects.
In single account mode registering new accounts doesn't make much sense, as just a single account is accessible anyway.

About refactoring the Manager, it's definitely too large, I'm planning to convert it to separate jobs, but that still needs more thinking.

**Reply by AsamK** [reply]
It's now in master (8f781c019f5f451c9d6323659bb248be335ad0e5)

**Reply by bublath** [reply]
Thanks. I'm currently looking into it. I worked on another (non-IT) project during the summer but now I probably can do some coding again.
First thing I noticed: It might make sense to implement the "isRegistered()" method with "false" for the DbusSignalControlImpl class, so the client can determine without an error which interface he is using (could be seen from the object path or doing an introspective as well, but this feels cleaner and easier).
I'll add a pull request once I did some more checking in case I have more change ideas. Let me know before you bump the version, so I can get that in :)

**Reply by bublath** [reply]
One thing I found is that providing a nonsense number (starting with + but e.g. +000000000000) is creating a rather cryptic reply:
NonSuccessfulResponseCodeException [400] Bad response: 400
I would expect an InvalidNumberException instead.

**Reply by bublath** [reply]
Correction, I just fetched the latest commits and now I get: 
Error Executing Method org.asamk.Signal.register: null

**Reply by bublath** [reply]
Number check works great with latest commit.
However I now fail to register. When I submit with incorrect or expired captcha, I get an error.
After I submitted a hopefully correct one, signal-cli freezes. Syslog shows in constant repeats:
`Aug 31 00:21:32 raspi400 signal-cli[23898]: WARN DaemonCommand - Receiving messages failed, retrying
Aug 31 00:21:32 raspi400 signal-cli[23898]: java.io.IOException: Connection closed!
Aug 31 00:21:32 raspi400 signal-cli[23898]: #011at org.whispersystems.signalservice.internal.websocket.WebSocketConnection.readRequest(WebSocketConnection.java:191)
Aug 31 00:21:32 raspi400 signal-cli[23898]: #011at org.whispersystems.signalservice.api.SignalWebSocket.readOrEmpty(SignalWebSocket.java:244)
Aug 31 00:21:32 raspi400 signal-cli[23898]: #011at org.asamk.signal.manager.Manager.receiveMessages(Manager.java:876)
Aug 31 00:21:32 raspi400 signal-cli[23898]: #011at org.asamk.signal.commands.DaemonCommand.lambda$run$1(DaemonCommand.java:133)
Aug 31 00:21:32 raspi400 signal-cli[23898]: #011at java.base/java.lang.Thread.run(Thread.java:829)
`

**Comment by bublath** [comment]
I believe I figured out the problem that caused a hang. My code simply had a glitch when a DBus call does not return data (void function). 
Got a successful call with a verification code and saw no error (not sure what caused the error mentioned above last time)
Calling it a day and hope to complete with implementing the verification tomorrow to see if the flow works all the way.

**Comment by bublath** [comment]
@AsamK I basically finished ny implementation and everrhing from register over captcha (fully automated on a Windows client thanks to the hint I got from @Florian7843 ) to verify is working. I would still appreciate to merge my small PR.
Great work. I'm still polishing my code a bit and plan to release a beta once the next version is bumped.
Now it would also be neat to think about providing apt packages that install with the dbus service in registration mode
